### PR TITLE
Auto generate backend key based on component name for azurerm

### PR DESCRIPTION
**Pull request by lukeorellana**
## what
When using the AzureRM backend type, automatically generate backend key based on component name if the backend key is not specified at the component level:
- Key is created in `{Global Backend Key value}/{component name}.terraform.tfstate` format. Places all component states in a storage account folder with the name of the global backend key value.
- If the key is specified at the component level, the backend key **is not** automatically generated.

**Before**:
```
terraform:
  backend_type: azurerm
  backend:
    azurerm:
      subscription_id: 88888-8888-8888-8888-8888888888
      resource_group_name: rg-terraform-state
      storage_account_name: staterraformstate
      container_name: dev-tfstate

components:
  terraform:
    "monitoring":
      backend:
          azurerm:
            key: dev.atmos/monitoring.terraform.tfstate
        
    "networking":
      backend:
          azurerm:
            key: dev.atmos/networking.terraform.tfstate
```

**After**:
```
terraform:
  backend_type: azurerm
  backend:
    azurerm:
      subscription_id: 88888-8888-8888-8888-8888888888
      resource_group_name: rg-terraform-state
      storage_account_name: staterraformstate
      container_name: dev-tfstate
      key: dev.atmos

components:
  terraform:
    "rg": {}

    "monitoring": {}

    "vault": {}
```
**backend.tf**:
```
{
  "terraform": {
    "backend": {
      "azurerm": {
        "subscription_id": "88888-8888-8888-8888-8888888888",
        "container_name": "dev-tfstate",
        "key": "dev.atmos/rg.terraform.tfstate",
        "resource_group_name": "rg-terraform-state",
        "storage_account_name": "staterraformstate"
      }
    }
  }
}
```

## why
The (azure backend) does not currently have a prefix option for workspaces. When using Atmos with Azure, users are forced to specify each backend key for every component.

## references
https://github.com/hashicorp/terraform/issues/28985

### CODE REVIEW

**Comment by aknysh** [comment]
@lukeorellana thanks for the PR, and thanks for trying out `atmos`.
The PR looks, good, would yuo please add an example of the backend here https://github.com/cloudposse/atmos/blob/master/examples/complete/stacks/globals/globals.yaml#L7 

```
backend_type: s3 # s3, remote, vault, static, azurerm, etc.
```

```
azurerm:
      subscription_id: 88888-8888-8888-8888-8888888888
      resource_group_name: rg-terraform-state
      storage_account_name: staterraformstate
      container_name: dev-tfstate
      key: dev.atmos
```

so other people would see the example on how to use it.

thanks!

**Review by aknysh — COMMENTED** [review]
@lukeorellana  thanks you, please see comments

**Comment by lukeorellana** [comment]
@aknysh Thanks! Pushed commit with backend example . Let me know if you would like me to add anything else.

File: pkg/stack/stack_processor.go [hunk 1]
```diff
@@ -478,6 +479,28 @@ func ProcessConfig(
...
+							}
+							if azurerm, ok2 := globalBackendSection["azurerm"].(map[interface{}]interface{}); ok2 {
+								baseKeyName = azurerm["key"].(string)
+							}
+							componentKeyName := strings.Replace(azureKeyPrefixComponent, "/", "-", -1)
+							finalComponentBackend["key"] = fmt.Sprintf("%s/%s.terraform.tfstate", baseKeyName, componentKeyName)
+
+						}
+					}
+
```
Comment (aknysh) [inline pkg/stack/stack_processor.go:501]:
```suggestion
```

File: pkg/stack/stack_processor.go [hunk 2]
```diff
@@ -478,6 +479,28 @@ func ProcessConfig(
 					}
 				}
 
+				// Check if component `backend` section has `key` for `azurerm` backend type
+				// If it does not, use the component name instead and format it with the global backend key name to auto generate a unique tf state key
+				// The backend state file will be formated like so: {global key name}/{component name}.terraform.tfstate
```
Comment (aknysh) [inline pkg/stack/stack_processor.go:484]:
```suggestion
				// The backend state file will be formatted like so: {global key name}/{component name}.terraform.tfstate
```

**Review by aknysh — APPROVED** [review]

**Comment by aknysh** [comment]
thanks @lukeorellana

**Pull request merged**